components/layout_2020/flow/mod.rs

```diff
@@ -486,23 +541,155 @@ fn layout_in_flow_replaced_block_level<'a>(
+        .maybe_percentage_relative_to(cbbs)
+        .auto_is(Length::zero);
+    let max_block_size = match max_box_size.block {
+        MaxSize::LengthPercentage(max_block_size) => {
+            max_block_size.maybe_percentage_relative_to(cbbs)
+        },
+        MaxSize::None => None,
+    };
+    let clamp = |inline_size, block_size| {
+        (
+            clamp_between_extremums(inline_size, min_inline_size, max_inline_size),
+            clamp_between_extremums(block_size, min_block_size, max_block_size),
+        )
+    };
+    // https://drafts.csswg.org/css2/visudet.html#min-max-widths
+    // https://drafts.csswg.org/css2/visudet.html#min-max-heights
     let (inline_size, block_size) = match (inline_size, block_size) {
         (LengthOrAuto::LengthPercentage(inline), LengthOrAuto::LengthPercentage(block)) => {
-            (inline, block)
+            clamp(inline, block)
         },
         (LengthOrAuto::LengthPercentage(inline), LengthOrAuto::Auto) => {
-            (inline, inline / intrinsic_ratio)
+            clamp(inline, inline / intrinsic_ratio)
         },
         (LengthOrAuto::Auto, LengthOrAuto::LengthPercentage(block)) => {
-            (block * intrinsic_ratio, block)
+            clamp(block * intrinsic_ratio, block)
+        },
+        (LengthOrAuto::Auto, LengthOrAuto::Auto) => {
+            enum Violation {
+                None,
+                Below(Length),
+                Above(Length),
+            }
+            let violation = |size, min_size, mut max_size: Option<Length>| {
+                if let Some(max) = max_size.as_mut() {
+                    max.max_assign(min_size);
+                }
+                if size < min_size {
+                    return Violation::Below(min_size);
+                }
+                match max_size {
+                    Some(max_size) if size > max_size => Violation::Above(max_size),
+                    _ => Violation::None,
+                }
+            };
+            match (
+                violation(intrinsic_size.inline, min_inline_size, max_inline_size),
+                violation(intrinsic_size.block, min_block_size, max_block_size),
+            ) {
+                // Row 1.
+                (Violation::None, Violation::None) => (intrinsic_size.inline, intrinsic_size.block),
+                // Row 2.
+                (Violation::Above(max_inline_size), Violation::None) => {
+                    let block_size = (max_inline_size / intrinsic_ratio).max(min_block_size);
+                    (max_inline_size, block_size)
+                },
+                // Row 3.
+                (Violation::Below(min_inline_size), Violation::None) => {
+                    let mut block_size = min_inline_size / intrinsic_ratio;
+                    if let Some(max_block_size) = max_block_size {
+                        if block_size > max_block_size {
+                            block_size = max_block_size;
+                        }
+                    }
```

Comment:
Nit: why not use `.min()` here? Or even for a function that handles the `Option`

Reply:
Something like:

```rust
fn maybe_clamp_at_max(x: Length, max_x: Option<Length>) -> Length {
    if let Some(max_x) = max_x {
        x.min(max_x)
    } else {
        x
    }
}